### ISSUE TITLE
Map InaSAFE's Metadata Properties to ISO 19115 

### ISSUE THREAD

**Issue by ismailsunni** [inasafe/inasafe#2538]
# Problem

## InaSAFE's metadata properties

We have mapped several properties to existing part in ISO 19115 metadata. We also couldn't find some attributes in the standard, so we put it under `supplementalInformation` tag.

Currently we have base metadata for the base class for the other metadata class (aggregation, exposure, hazard, and impact). Here are the properties of each class

### Base Metadata
https://github.com/inasafe/inasafe/blob/iso_keywords_2/safe/metadata/base_metadata.py#L62
- [x] organisation
- [x] email
- [x] date
- [x] abstract
- [x] title
- [x] license
- [x] url
- [ ] report
- [ ] layer_purpose
- [ ] layer_mode
- [ ] layer_geometry
- [ ] keyword_version
- [ ] scale
- [ ] source
- [ ] datatype
- [ ] multipart_polygon
- [ ] resolution

### Aggregation Metadata
https://github.com/ismailsunni/inasafe/blob/iso_keywords_2/safe/metadata/aggregation_layer_metadata.py#L31
- [ ] aggregation attribute
- [ ] adult ratio attribute
- [ ] adult ratio default
- [ ] elderly ratio attribute
- [ ] elderly ratio default
- [ ] female ratio attribute
- [ ] female ratio default
- [ ] youth ratio attribute
- [ ] youth ratio default

### Exposure Metadata
https://github.com/ismailsunni/inasafe/blob/iso_keywords_2/safe/metadata/exposure_layer_metadata.py#L31
- [ ] exposure
- [ ] exposure_unit
- [ ] structure_class_field
- [ ] road_class_field
- [ ] field
- [ ] area_population_field
- [ ] area_name_field

### Hazard Metadata
https://github.com/ismailsunni/inasafe/blob/iso_keywords_2/safe/metadata/hazard_layer_metadata.py#L31
- [ ] hazard
- [ ] hazard_category
- [ ] continuous_hazard_unit
- [ ] vector_hazard_classification
- [ ] raster_hazard_classification
- [ ] field
- [ ] value_map
- [ ] volcano_name_field
- [ ] prob_fatality_mag

### Impact Metadata 
https://github.com/ismailsunni/inasafe/blob/iso_keywords_2/safe/metadata/impact_layer_metadata.py#L53
- [ ] elapsed_time
- [ ] hazard_title
- [ ] postprocessing_report
- [ ] exposure_title
- [ ] legend_title
- [ ] legend_notes
- [ ] exposure_source
- [ ] map_title
- [ ] legend_units
- [ ] impact_summary
- [ ] user
- [ ] host_name
- [ ] time_stamp
- [ ] hazard_source
- [ ] target_field
- [ ] impact_table
- [ ] statistics_classes
- [ ] statistics_type

Note: properties with checked check box has already put in standard ISO 19115 location.

## Property Type

We also "invent" some data type to be stored in the ISO:
From ISO 19115:
- gco:CharacterString
- gco:Date
- gmd:URL

Our "invention":
- gco:Dictionary
- gco:Integer
- gco:Boolean
- gco:Float
- gco:List
- gco:Tuple

We have those list, and most of them are not mapped correctly or can't be mapped to standard ISO 19115 properties. We wish we can map as many as possible.

# Expected Result
We map as many as possible our metadata properties to ISO 19115 standard properties so that the metadata can be used in another system seamlessly.

cc @timlinux @tomkralidis

**Comment by Gustry** [comment]
Can we have an update on this ticket @ismailsunni ? Thanks

**Comment by ismailsunni** [comment]
Update:
- It's not easy to map our metadata to the existing one. Mostly, everything is not map-able.
- We use non-standard way to put our metadata. That makes it not compatible. We need to make it as a string, see https://github.com/inasafe/inasafe/issues/4251

**Comment by Gustry** [comment]
Yes, I think it's not easy. Thanks for the update.

But I think we should focus on this issue. QGIS 3 is coming with the QgsLayerMetadata class and we need to figure out how we will store our metadata in this class.

Maybe first we should clean this list above. For instance, in the aggregation layer, they are all in the `inasafe_fields` and `inasafe_default_values`.

I guess it's ok to have not map-able items such as `inasafe_fields`, `inasafe_default_values`, `extra_keywords` because they are very specific to InaSAFE.

What do you mean by `non-standard way`? These `gco:Dictionary` etc?

The feature freeze for the QGIS Internal schema is in a few weeks.

**Comment by ismailsunni** [comment]
@Gustry 
Yes, the list above is not updated. I haven't looked at this ticket before you ping me here.

The other properties are also very specific to InaSAFE, like purpose, layer mode, exposure unit, hazard unit, hazard classification etc.

> What do you mean by non-standard way? These gco:Dictionary etc?

We store it as dictionary, but it should be a string if we want to be compatible with ISO19115.

> The feature freeze for the QGIS Internal schema is in a few weeks.

I think, as long as the QGIS Internal schema has something like extensible or suplementary_data tag, we can add our metadata to it. I am not expert, but I think there are many use case for having very specific metadata in the other part of the world.

**Comment by Gustry** [comment]
> We store it as dictionary, but it should be a string if we want to be compatible with ISO19115.

One string only? I just understood the issue sorry ;-)

> I think, as long as the QGIS Internal schema has something like extensible or suplementary_data tag, we can add our metadata to it.

Not really for now and we would like to avoid showing an unreadable JSON string in the QGIS metadata viewer representing InaSAFE data. Ideally, we would like to provide the same viewer as we have in the dock. To be able to render dictionaries and nested dictionaries for instance like we do in the dock about classification, value map ... 

> I am not expert, but I think there are many use case for having very specific metadata in the other part of the world.

Yes!, I agree. I don't see any reason to expose `inasafe_fields`, `inasafe_default_values`, `value_map` and so on ... These are useless outside of InaSAFE.

**Comment by ismailsunni** [comment]
> Yes!, I agree. I don't see any reason to expose inasafe_fields, inasafe_default_values, value_map and so on ... These are useless outside of InaSAFE.

No, I mean, I believe that there are people who need to store very specific metadata like us. I wonder, how do they solve the problem.
